// @staratlas/galactic-marketplace/src/instructions/addFeeExemption.ts
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram, MarketVars } from '..';
import { FeeReduction } from '../state/feeReduction';

/**
 * Returns an instruction which flags an account as fee exempt. Update authority must match marketVarsAccount.
 * @param program - Marketplace program
 * @param updateAuthorityAccount - Public key of registered update authority
 * @param feeExemptTarget - Target account to exempt from fees
 * @returns InstructionReturn
 */
export function addFeeExpemption(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  feeExemptTarget: PublicKey,
): InstructionReturn {
  const discount = new BN(1_000_000);
  const feeExemptAccount = FeeReduction.findAddress(
    feeExemptTarget,
    program,
  )[0];
  const marketVarsAccount = MarketVars.findAddress(program)[0];
  return async (funder) => [
    {
      instruction: await program.methods
        .addFeeExemption(discount)
        .accountsStrict({
          updateAuthorityMaster: updateAuthorityAccount.publicKey(),
          funder: funder.publicKey(),
          feeExemptTarget,
          feeExemptAccount,
          marketVarsAccount,
          systemProgram: SystemProgram.programId,
        })
        .instruction(),
      signers: [updateAuthorityAccount, funder],
    },
  ];
}
