// @staratlas/galactic-marketplace/src/instructions/cancelOrder.ts
import {
  Account,
  TOKEN_2022_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
} from '@solana/spl-token';
import {
  AccountMeta,
  Connection,
  Keypair,
  PublicKey,
  SYSVAR_INSTRUCTIONS_PUBKEY,
} from '@solana/web3.js';
import { InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars, OpenOrdersCounter, Order } from '../state';
import {
  findExtraAccountMetaListAddress,
  getOrderSide,
  getTokenAccount,
} from '../util';
import { createOpenOrdersCounter } from './createOpenOrdersCounter';

/**
 * Cancels Order with Solana connection
 * Takes a connection to fetch necessary accounts - it is preferred to use `cancelAccount`
 * if account data can be provided without an RPC connection.
 * @param connection - Connection
 * @param orderInitializer - Order Initializer Key
 * @param orderAccount - Order Account
 * @param program - Galactic Marketplace Program
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export async function cancelOrderWithConnection(
  connection: Connection,
  orderInitializer: PublicKey,
  orderAccount: Order,
  program: GalacticMarketplaceIDLProgram,
  transferHookProgramId?: PublicKey,
): Promise<InstructionReturn[]> {
  const instructionArray: InstructionReturn[] = [];
  const orderAccountInfo = orderAccount.data;
  const orderSide = getOrderSide(orderAccount);
  const depositMint =
    orderSide === 'SellSide'
      ? orderAccountInfo.assetMint
      : orderAccountInfo.currencyMint;
  // Get initializer deposit token account, pushing creation instruction if necessary
  let initializerDepositTokenAccount: PublicKey;
  const response = await getTokenAccount(
    connection,
    orderInitializer,
    depositMint,
  );
  const tokenAccount: PublicKey | Keypair = response.tokenAccount;
  if ('createInstruction' in response) {
    instructionArray.push(response.createInstruction);
  }
  if (tokenAccount instanceof Keypair) {
    initializerDepositTokenAccount = tokenAccount.publicKey;
  } else {
    initializerDepositTokenAccount = tokenAccount;
  }

  // Get open orders counter, pushing creation instruction if necessary
  const [openOrdersCounterAddress] = OpenOrdersCounter.findAddress(
    program,
    orderInitializer,
    depositMint,
  );
  const orderCounter = await connection.getAccountInfo(
    openOrdersCounterAddress,
  );
  if (orderCounter === null) {
    const createOpenOrderCounterInstruction = await createOpenOrdersCounter(
      program,
      orderInitializer,
      depositMint,
    );
    instructionArray.push(createOpenOrderCounterInstruction);
  }
  const cancelInstruction = cancelOrder(
    program,
    orderAccount,
    orderInitializer,
    { key: initializerDepositTokenAccount },
    { key: openOrdersCounterAddress },
    orderSide === 'SellSide' ? transferHookProgramId : undefined,
  );
  instructionArray.push(cancelInstruction);
  return instructionArray;
}

/**
 * Cancels Order
 * @param program -Galactic Marketplace Program
 * @param orderAccount - Order Account
 * @param orderInitializer - Order Initializer Key
 * @param initializerDepositTokenAccount - Deposit Token Account
 * @param openOrdersCounter - OpenOrdersCounter Account/Key
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export function cancelOrder(
  program: GalacticMarketplaceIDLProgram,
  orderAccount: Order,
  orderInitializer: PublicKey,
  initializerDepositTokenAccount: { account: Account } | { key: PublicKey },
  openOrdersCounter:
    | { openOrdersCounter: OpenOrdersCounter }
    | { key: PublicKey },
  transferHookProgramId?: PublicKey,
): InstructionReturn {
  // Pass in order account and fetch data or pass in `Order`?
  const orderAccountInfo = orderAccount.data;
  const orderSide = getOrderSide(orderAccount);
  const depositMint =
    orderSide === 'SellSide'
      ? orderAccountInfo.assetMint
      : orderAccountInfo.currencyMint;
  return cancelOrderBareBones(
    program,
    orderAccount.key,
    orderSide,
    depositMint,
    orderInitializer,
    'key' in initializerDepositTokenAccount
      ? initializerDepositTokenAccount.key
      : initializerDepositTokenAccount.account.address,
    'key' in openOrdersCounter
      ? openOrdersCounter.key
      : openOrdersCounter.openOrdersCounter.key,
    transferHookProgramId,
  );
}

/**
 * Cancels Order (without requiring an Order account)
 * @param program -Galactic Marketplace Program
 * @param orderAccount - Order Account address
 * @param orderSide - the type of order (either buy or sell)
 * @param depositMint - the mint of the token that was deposited when the order was made
 * @param orderInitializer - Order Initializer Key
 * @param initializerDepositTokenAccount - Deposit Token Account
 * @param openOrdersCounter - OpenOrdersCounter Account/Key
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export function cancelOrderBareBones(
  program: GalacticMarketplaceIDLProgram,
  orderAccount: PublicKey,
  orderSide: 'BuySide' | 'SellSide',
  depositMint: PublicKey,
  orderInitializer: PublicKey,
  initializerDepositTokenAccount: PublicKey,
  openOrdersCounter: PublicKey,
  transferHookProgramId?: PublicKey,
): InstructionReturn {
  const [orderVaultAccount] = Order.getOrderVaultAddress(
    program,
    orderInitializer,
    depositMint,
  );

  const [orderVaultAuthority] = Order.getOrderVaultAuthorityAddress(
    program,
    orderInitializer,
  );

  const extraAccounts: AccountMeta[] =
    orderSide === 'SellSide' && transferHookProgramId
      ? [
          {
            pubkey: findExtraAccountMetaListAddress(
              transferHookProgramId,
              depositMint,
            )[0],
            isSigner: false,
            isWritable: false,
          },
          {
            pubkey: transferHookProgramId,
            isSigner: false,
            isWritable: false,
          },
          {
            pubkey: TOKEN_2022_PROGRAM_ID,
            isSigner: false,
            isWritable: false,
          },
          {
            pubkey: SYSVAR_INSTRUCTIONS_PUBKEY,
            isSigner: false,
            isWritable: false,
          },
        ]
      : [
          {
            pubkey: TOKEN_PROGRAM_ID,
            isSigner: false,
            isWritable: false,
          },
        ];

  return async (funder) => ({
    instruction: await program.methods
      .processCancel()
      .accountsStrict({
        marketVarsAccount: MarketVars.findAddress(program)[0],
        signer: funder.publicKey(),
        orderInitializer,
        depositMint,
        initializerDepositTokenAccount,
        orderVaultAccount,
        orderVaultAuthority,
        orderAccount,
        openOrdersCounter,
      })
      .remainingAccounts(extraAccounts)
      .instruction(),
    signers: [funder],
  });
}
