// @staratlas/galactic-marketplace/src/instructions/createOpenOrdersCounter.ts
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { OpenOrdersCounter } from '../state';

/**
 * Creates OpenOrdersCounter Account
 * @param program - Galactic Marketplace Program
 * @param user - User Key
 * @param depositMint - Deposite Mint key
 * @returns InstructionReturn
 */
export function createOpenOrdersCounter(
  program: GalacticMarketplaceIDLProgram,
  user: PublicKey,
  depositMint: PublicKey,
): InstructionReturn {
  return async (funder) => ({
    instruction: await program.methods
      .initializeOpenOrdersCounter()
      .accountsStrict({
        payer: funder.publicKey(),
        user,
        openOrdersCounter: OpenOrdersCounter.findAddress(
          program,
          user,
          depositMint,
        )[0],
        depositMint,
        systemProgram: SystemProgram.programId,
      })
      .instruction(),
    signers: [funder],
  });
}
