// @staratlas/galactic-marketplace/src/instructions/deregisterCurrency.ts
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars, RegisteredCurrency } from '../state';

/**
 * Deregisters a Currency Mint
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - Update Authority
 * @param currencyMint - Currency Mint to be disabled
 * @returns InstructionReturn
 */
export function deregisterCurrency(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  currencyMint: PublicKey,
): InstructionReturn {
  return async () => ({
    instruction: await program.methods
      .deregisterCurrency()
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        registeredCurrency: RegisteredCurrency.findAddress(
          program,
          currencyMint,
        )[0],
        currencyMint,
        systemProgram: SystemProgram.programId,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}
