// @staratlas/galactic-marketplace/src/instructions/exchangeOrder.ts
import { TOKEN_2022_PROGRAM_ID, TOKEN_PROGRAM_ID } from '@solana/spl-token';
import {
  AccountMeta,
  Connection,
  Keypair,
  PublicKey,
  SYSVAR_INSTRUCTIONS_PUBKEY,
} from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { OpenOrdersCounter, Order, RegisteredCurrency } from '../state';
import {
  findExtraAccountMetaListAddress,
  getOrderSide,
  getTokenAccount,
} from '../util';

/**
 * Executes an exchange with Solana connection
 * Takes a connection to fetch necessary accounts - it is preferred to use `exchangeAccount`
 * @param connection - Connection
 * @param orderAccount - Order Account
 * @param orderTaker - Order taker Account
 * @param orderTakerDepositTokenAccount - Taker deposit Token Account
 * @param purchaseQty - Purchase quantity
 * @param expectedPrice - Expected Price
 * @param saVault - Star Atlas vault account
 * @param stakingProgramId - Staking program ID
 * @param registeredStake - Registered Stake key
 * @param stakingAccount - Staking Account
 * @param program - GalacticMarketplace Program
 * @param extraAccounts - Extra accounts, used for buddy link
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export async function exchangeOrderWithConnection(
  connection: Connection,
  orderAccount: Order,
  orderTaker: AsyncSigner,
  orderTakerDepositTokenAccount: PublicKey,
  purchaseQty: number,
  expectedPrice: BN,
  saVault: PublicKey,
  stakingProgramId: PublicKey,
  registeredStake: PublicKey,
  stakingAccount: PublicKey,
  program: GalacticMarketplaceIDLProgram,
  extraAccounts: AccountMeta[] = [],
  transferHookProgramId?: PublicKey,
): Promise<InstructionReturn[]> {
  const instructions: InstructionReturn[] = [];
  const orderSide = getOrderSide(orderAccount);
  const orderInitializer = orderAccount.data.orderInitializerPubkey;
  const initializerDepositMint =
    orderSide === 'SellSide'
      ? orderAccount.data.assetMint
      : orderAccount.data.currencyMint;
  const initializerReceiveMint =
    orderSide === 'SellSide'
      ? orderAccount.data.currencyMint
      : orderAccount.data.assetMint;

  // Find initializer deposit account
  let initializerDepositTokenAccount: PublicKey;
  let response = await getTokenAccount(
    connection,
    orderInitializer,
    initializerDepositMint,
    orderTaker.publicKey(),
  );
  if ('createInstruction' in response) {
    instructions.push(response.createInstruction);
  }
  if (response.tokenAccount instanceof Keypair) {
    initializerDepositTokenAccount = response.tokenAccount.publicKey;
  } else {
    initializerDepositTokenAccount = response.tokenAccount;
  }
  // Find initializer receive account
  let initializerReceiveTokenAccount: PublicKey;
  response = await getTokenAccount(
    connection,
    orderInitializer,
    initializerReceiveMint,
    orderTaker.publicKey(),
  );
  if ('createInstruction' in response) {
    instructions.push(response.createInstruction);
  }
  if (response.tokenAccount instanceof Keypair) {
    initializerReceiveTokenAccount = response.tokenAccount.publicKey;
  } else {
    initializerReceiveTokenAccount = response.tokenAccount;
  }

  // Find order taker receive account
  let orderTakerReceiveTokenAccount: PublicKey;
  response = await getTokenAccount(
    connection,
    orderTaker.publicKey(),
    initializerDepositMint,
  );
  if ('createInstruction' in response) {
    instructions.push(response.createInstruction);
  }
  if (response.tokenAccount instanceof Keypair) {
    orderTakerReceiveTokenAccount = response.tokenAccount.publicKey;
  } else {
    orderTakerReceiveTokenAccount = response.tokenAccount;
  }

  const exchangeInstruction = exchangeOrder(
    program,
    orderAccount,
    orderTaker,
    purchaseQty,
    expectedPrice,
    initializerDepositTokenAccount,
    initializerReceiveTokenAccount,
    orderTakerDepositTokenAccount,
    orderTakerReceiveTokenAccount,
    saVault,
    stakingProgramId,
    registeredStake,
    stakingAccount,
    extraAccounts,
    transferHookProgramId,
  );
  instructions.push(exchangeInstruction);
  return instructions;
}

/**
 * Get the address of a Fee Exemption account
 * @param program - GalacticMarketplace Program
 * @param seller - the seller
 * @returns the PDA and bump
 */
export function getFeeReduction(
  program: GalacticMarketplaceIDLProgram,
  seller: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('fee_exempt'), seller.toBuffer()],
    program.programId,
  );
}

/**
 * Executes an exchange
 * @param program - GalacticMarketplace Program
 * @param orderAccount - the order account
 * @param orderTaker - the order take
 * @param purchaseQty - the purchase quantity in base units
 * @param expectedPrice - the expected price in base units
 * @param initializerDepositTokenAccount - the initializer deposit token account
 * @param initializerReceiveTokenAccount - the initializer receiving token account
 * @param orderTakerDepositTokenAccount - the order taker deposit token account
 * @param orderTakerReceiveTokenAccount - the order taker receiving token account
 * @param saVault - Star Atlas vault account - must match account in registered currency
 * @param stakingProgramId - the ATLAS staking program
 * @param registeredStake - Atlas `RegisteredStake` Account
 * @param stakingAccount - User's Atlas staking account
 * @param extraAccounts - Extra accounts, used for buddy link
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export function exchangeOrder(
  program: GalacticMarketplaceIDLProgram,
  orderAccount: Order,
  orderTaker: AsyncSigner,
  purchaseQty: number,
  expectedPrice: BN,
  initializerDepositTokenAccount: PublicKey,
  initializerReceiveTokenAccount: PublicKey,
  orderTakerDepositTokenAccount: PublicKey,
  orderTakerReceiveTokenAccount: PublicKey,
  saVault: PublicKey,
  stakingProgramId: PublicKey,
  registeredStake: PublicKey,
  stakingAccount: PublicKey,
  extraAccounts: AccountMeta[] = [],
  transferHookProgramId?: PublicKey,
): InstructionReturn {
  const orderSide = getOrderSide(orderAccount);
  const initializerDepositMint =
    orderSide === 'SellSide'
      ? orderAccount.data.assetMint
      : orderAccount.data.currencyMint;
  const orderInitializer = orderAccount.data.orderInitializerPubkey;

  return exchangeOrderBareBones(
    program,
    orderAccount.key,
    orderSide,
    orderAccount.data.assetMint,
    orderAccount.data.currencyMint,
    orderTaker,
    purchaseQty,
    expectedPrice,
    orderInitializer,
    initializerDepositMint,
    initializerDepositTokenAccount,
    initializerReceiveTokenAccount,
    orderTakerDepositTokenAccount,
    orderTakerReceiveTokenAccount,
    saVault,
    stakingProgramId,
    registeredStake,
    stakingAccount,
    extraAccounts,
    transferHookProgramId,
  );
}

/**
 * Executes an exchange (without requiring an Order account)
 * @param program - GalacticMarketplace Program
 * @param orderAccount - the order account
 * @param orderSide - the type of order (either buy or sell)
 * @param assetMint - the order's asset mint address
 * @param currencyMint - the order's currency mint address
 * @param orderTaker - the order take
 * @param purchaseQty - the purchase quantity in base units
 * @param expectedPrice - the expected price in base units
 * @param orderInitializer - the order initializer's address
 * @param initializerDepositMint - the order initializer's deposit mint
 * @param initializerDepositTokenAccount - the initializer deposit token account
 * @param initializerReceiveTokenAccount - the initializer receiving token account
 * @param orderTakerDepositTokenAccount - the order taker deposit token account
 * @param orderTakerReceiveTokenAccount - the order taker receiving token account
 * @param saVault - Star Atlas vault account - must match account in registered currency
 * @param stakingProgramId - the ATLAS staking program
 * @param registeredStake - Atlas `RegisteredStake` Account
 * @param stakingAccount - User's Atlas staking account
 * @param extraAccounts - Extra accounts, used for buddy link
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export function exchangeOrderBareBones(
  program: GalacticMarketplaceIDLProgram,
  orderAccount: PublicKey,
  orderSide: 'BuySide' | 'SellSide',
  assetMint: PublicKey,
  currencyMint: PublicKey,
  orderTaker: AsyncSigner,
  purchaseQty: number,
  expectedPrice: BN,
  orderInitializer: PublicKey,
  initializerDepositMint: PublicKey,
  initializerDepositTokenAccount: PublicKey,
  initializerReceiveTokenAccount: PublicKey,
  orderTakerDepositTokenAccount: PublicKey,
  orderTakerReceiveTokenAccount: PublicKey,
  saVault: PublicKey,
  stakingProgramId: PublicKey,
  registeredStake: PublicKey,
  stakingAccount: PublicKey,
  extraAccounts: AccountMeta[] = [],
  transferHookProgramId?: PublicKey,
): InstructionReturn {
  const seller =
    orderSide === 'SellSide' ? orderInitializer : orderTaker.publicKey();
  const [feeReduction] = getFeeReduction(program, seller);

  if (transferHookProgramId) {
    extraAccounts.push(
      ...[
        {
          pubkey: findExtraAccountMetaListAddress(
            transferHookProgramId,
            assetMint,
          )[0],
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: transferHookProgramId,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: TOKEN_2022_PROGRAM_ID,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: SYSVAR_INSTRUCTIONS_PUBKEY,
          isSigner: false,
          isWritable: false,
        },
      ],
    );
  }

  return async () => ({
    instruction: await program.methods
      .processExchange(new BN(purchaseQty), expectedPrice, seller)
      .accountsStrict({
        openOrdersCounter: OpenOrdersCounter.findAddress(
          program,
          orderInitializer,
          initializerDepositMint,
        )[0],
        orderAccount,
        orderInitializer,
        orderVaultAccount: Order.getOrderVaultAddress(
          program,
          orderInitializer,
          initializerDepositMint,
        )[0],
        orderVaultAuthority: Order.getOrderVaultAuthorityAddress(
          program,
          orderInitializer,
        )[0],
        initializerDepositTokenAccount,
        initializerReceiveTokenAccount,
        registeredCurrency: RegisteredCurrency.findAddress(
          program,
          currencyMint,
        )[0],
        tokenProgram: TOKEN_PROGRAM_ID,
        orderTaker: orderTaker.publicKey(),
        orderTakerDepositTokenAccount,
        orderTakerReceiveTokenAccount,
        currencyMint,
        assetMint,
        saVault,
        atlasStaking: stakingProgramId,
        registeredStake,
        stakingAccount,
        feeReduction,
      })
      .remainingAccounts(extraAccounts)
      .instruction(),
    signers: [orderTaker],
  });
}
