// @staratlas/galactic-marketplace/src/instructions/initializeMarket.ts
import { SystemProgram } from '@solana/web3.js';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars } from '../state';

/**
 * Initialize the marketplace program
 * This should be done once when the program is newly deployed to set the update authority
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the new market
 * @returns InstructionReturn
 */
export function initializeMarket(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
): InstructionReturn {
  return async () => ({
    instruction: await program.methods
      .initializeMarketplace()
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        systemProgram: SystemProgram.programId,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}
