// @staratlas/galactic-marketplace/src/instructions/processOrder.ts
import { TOKEN_2022_PROGRAM_ID, TOKEN_PROGRAM_ID } from '@solana/spl-token';
import {
  AccountMeta,
  Connection,
  Keypair,
  PublicKey,
  SystemProgram,
  SYSVAR_INSTRUCTIONS_PUBKEY,
  SYSVAR_RENT_PUBKEY,
} from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import {
  MarketVars,
  OpenOrdersCounter,
  Order,
  RegisteredCurrency,
} from '../state';
import {
  createOrderAccountInstruction,
  findExtraAccountMetaListAddress,
  getTokenAccount,
} from '../util';
import { createOpenOrdersCounter } from './createOpenOrdersCounter';

/**
 * Creates a new buy/sell order with Solana connection
 * Takes a connection to fetch necessary accounts - it is preferred to use `initializeOrder`
 * if account data can be provided without an RPC connection.
 * @param connection - the Solana connection
 * @param side - Order side
 * @param orderInitializer - AsyncSigner creating the order
 * @param price - price In base units
 * @param originationQty - Quantity to sell or buy
 * @param itemMint - the mint of the item being sold or bought
 * @param currencyMint - The currency, must be registered already with the program
 * @param program - GalacticMarketplace Program
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export async function initializeOrderWithConnection(
  connection: Connection,
  side: 'sell' | 'buy',
  orderInitializer: AsyncSigner,
  price: BN,
  originationQty: number,
  itemMint: PublicKey,
  currencyMint: PublicKey,
  program: GalacticMarketplaceIDLProgram,
  transferHookProgramId?: PublicKey,
): Promise<{
  orderAccount: PublicKey;
  instructions: InstructionReturn[];
}> {
  const instructionArray: InstructionReturn[] = [];

  const depositMint = side === 'sell' ? itemMint : currencyMint;
  const receiveMint = side === 'sell' ? currencyMint : itemMint;

  /**
   * This makes the assumption that the deposit token account ATA exists
   * Which it should, if the user is trying to transact the deposit token
   */
  const { tokenAccount: depositTokenAccount } = await getTokenAccount(
    connection,
    orderInitializer.publicKey(),
    depositMint,
    orderInitializer.publicKey(),
    undefined,
    transferHookProgramId ? TOKEN_2022_PROGRAM_ID : TOKEN_PROGRAM_ID,
  );
  const initializerDepositTokenAccountAddress: PublicKey =
    'publicKey' in depositTokenAccount
      ? depositTokenAccount.publicKey
      : depositTokenAccount;

  let initializerReceiveTokenAccount: PublicKey;

  const receiveAtaResponse = await getTokenAccount(
    connection,
    orderInitializer.publicKey(),
    receiveMint,
    orderInitializer.publicKey(),
    undefined,
    transferHookProgramId ? TOKEN_2022_PROGRAM_ID : TOKEN_PROGRAM_ID,
  );
  const receiveTokenAccount: PublicKey | Keypair =
    receiveAtaResponse.tokenAccount;

  if ('createInstruction' in receiveAtaResponse) {
    instructionArray.push(receiveAtaResponse.createInstruction);
  }

  if (receiveTokenAccount instanceof Keypair) {
    initializerReceiveTokenAccount = receiveTokenAccount.publicKey;
  } else {
    initializerReceiveTokenAccount = receiveTokenAccount;
  }
  // Get open order counter, initializing if necessary
  const [openOrdersCounterAddress] = OpenOrdersCounter.findAddress(
    program,
    orderInitializer.publicKey(),
    depositMint,
  );
  const orderCounter = await connection.getAccountInfo(
    openOrdersCounterAddress,
  );

  if (orderCounter === null) {
    const createOpenOrderCounterInstruction = createOpenOrdersCounter(
      program,
      orderInitializer.publicKey(),
      depositMint,
    );
    instructionArray.push(createOpenOrderCounterInstruction);
  }

  // Generate order account
  const orderAccount = Keypair.generate();
  const orderAccountInitializeInstruction = await createOrderAccountInstruction(
    connection,
    program,
    orderAccount,
  );

  instructionArray.push(orderAccountInitializeInstruction);

  const orderInstruction = initializeOrder(
    side,
    program,
    orderAccount.publicKey,
    price,
    originationQty,
    initializerDepositTokenAccountAddress,
    initializerReceiveTokenAccount,
    depositMint,
    receiveMint,
    currencyMint,
    orderInitializer,
    openOrdersCounterAddress,
    transferHookProgramId,
  );
  instructionArray.push(orderInstruction);

  return {
    orderAccount: orderAccount.publicKey,
    instructions: instructionArray,
  };
}

/**
 * Creates a new buy/sell order
 * @param side - Order side
 * @param program - GalacticMarketplace Program Anchor program
 * @param orderAccount - The publicKey of the order account created for this order, must exist already
 * @param price - The price in base units
 * @param originationQty - Quantity to sell or buy in base units
 * @param initializerDepositTokenAccount - depositMint token account
 * @param initializerReceiveTokenAccount - receiveMint token account
 * @param depositMint - What the user is depositing, e.g. currency if buying, item if selling
 * @param receiveMint - What the user wants in exchange, e.g. item if buying, currency if selling
 * @param currencyMint - The currency, must be registered already in the program
 * @param orderInitializer - AsyncSigner creating the order
 * @param openOrdersCounter - Should exist for the depositMint
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export function initializeOrder(
  side: 'sell' | 'buy',
  program: GalacticMarketplaceIDLProgram,
  orderAccount: PublicKey,
  price: BN,
  originationQty: number,
  initializerDepositTokenAccount: PublicKey,
  initializerReceiveTokenAccount: PublicKey,
  depositMint: PublicKey,
  receiveMint: PublicKey,
  currencyMint: PublicKey,
  orderInitializer: AsyncSigner,
  openOrdersCounter: PublicKey,
  transferHookProgramId?: PublicKey,
): InstructionReturn {
  const orderVaultAccount = Order.getOrderVaultAddress(
    program,
    orderInitializer.publicKey(),
    depositMint,
  )[0];
  const orderVaultAuthority = Order.getOrderVaultAuthorityAddress(
    program,
    orderInitializer.publicKey(),
  )[0];
  const marketVarsAccount = MarketVars.findAddress(program)[0];
  const registeredCurrency = RegisteredCurrency.findAddress(
    program,
    currencyMint,
  )[0];

  const programMethod =
    side === 'buy'
      ? program.methods.processInitializeBuy
      : program.methods.processInitializeSell;

  const extraAccounts: AccountMeta[] = transferHookProgramId
    ? [
        {
          pubkey: findExtraAccountMetaListAddress(
            transferHookProgramId,
            side === 'buy' ? receiveMint : depositMint,
          )[0],
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: transferHookProgramId,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: TOKEN_2022_PROGRAM_ID,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: SYSVAR_INSTRUCTIONS_PUBKEY,
          isSigner: false,
          isWritable: false,
        },
      ]
    : [];

  return async () => ({
    instruction: await programMethod(price, new BN(originationQty))
      .accountsStrict({
        orderVaultAccount,
        orderVaultAuthority,
        orderInitializer: orderInitializer.publicKey(),
        marketVarsAccount,
        depositMint,
        receiveMint,
        initializerDepositTokenAccount,
        initializerReceiveTokenAccount,
        orderAccount,
        registeredCurrency,
        openOrdersCounter,
        systemProgram: SystemProgram.programId,
        rent: SYSVAR_RENT_PUBKEY,
        tokenProgram: TOKEN_PROGRAM_ID,
      })
      .remainingAccounts(extraAccounts)
      .instruction(),
    signers: [orderInitializer],
  });
}

/**
 * Creates a new buy/sell order using a PDA for the order account
 * @param side - Order side
 * @param program - GalacticMarketplace Program Anchor program
 * @param price - The price in base units
 * @param originationQty - Quantity to sell or buy in base units
 * @param initializerDepositTokenAccount - depositMint token account
 * @param initializerReceiveTokenAccount - receiveMint token account
 * @param depositMint - What the user is depositing, e.g. currency if buying, item if selling
 * @param receiveMint - What the user wants in exchange, e.g. item if buying, currency if selling
 * @param currencyMint - The currency, must be registered already in the program
 * @param orderInitializer - AsyncSigner creating the order
 * @param openOrdersCounter - Should exist for the depositMint
 * @param transferHookProgramId - The optional transfer hook program id
 * @returns InstructionReturn
 */
export function initializeOrderPda(
  side: 'sell' | 'buy',
  program: GalacticMarketplaceIDLProgram,
  price: BN,
  originationQty: number,
  initializerDepositTokenAccount: PublicKey,
  initializerReceiveTokenAccount: PublicKey,
  depositMint: PublicKey,
  receiveMint: PublicKey,
  currencyMint: PublicKey,
  orderInitializer: AsyncSigner,
  openOrdersCounter: PublicKey,
  transferHookProgramId?: PublicKey,
): InstructionReturn {
  const orderVaultAccount = Order.getOrderVaultAddress(
    program,
    orderInitializer.publicKey(),
    depositMint,
  )[0];
  const orderVaultAuthority = Order.getOrderVaultAuthorityAddress(
    program,
    orderInitializer.publicKey(),
  )[0];
  const marketVarsAccount = MarketVars.findAddress(program)[0];
  const registeredCurrency = RegisteredCurrency.findAddress(
    program,
    currencyMint,
  )[0];
  const orderAccountPubkey = PublicKey.unique();
  const [orderAccountPda] = PublicKey.findProgramAddressSync(
    [orderAccountPubkey.toBuffer()],
    program.programId,
  );

  const programMethod =
    side === 'buy'
      ? program.methods.processInitializeBuyPda
      : program.methods.processInitializeSellPda;

  const extraAccounts: AccountMeta[] = transferHookProgramId
    ? [
        {
          pubkey: findExtraAccountMetaListAddress(
            transferHookProgramId,
            depositMint,
          )[0],
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: transferHookProgramId,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: TOKEN_2022_PROGRAM_ID,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: SYSVAR_INSTRUCTIONS_PUBKEY,
          isSigner: false,
          isWritable: false,
        },
      ]
    : [];

  return async (funder) => ({
    instruction: await programMethod(
      price,
      new BN(originationQty),
      orderAccountPubkey,
    )
      .accountsStrict({
        payer: funder.publicKey(),
        orderVaultAccount,
        orderVaultAuthority,
        orderInitializer: orderInitializer.publicKey(),
        marketVarsAccount,
        depositMint,
        receiveMint,
        initializerDepositTokenAccount,
        initializerReceiveTokenAccount,
        orderAccount: orderAccountPda,
        registeredCurrency,
        openOrdersCounter,
        systemProgram: SystemProgram.programId,
        rent: SYSVAR_RENT_PUBKEY,
        tokenProgram: TOKEN_PROGRAM_ID,
      })
      .remainingAccounts(extraAccounts)
      .instruction(),
    signers: [orderInitializer],
  });
}
