// @staratlas/galactic-marketplace/src/instructions/registerCurrency.ts
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars, RegisteredCurrency } from '../state';

/**
 * Creates a new registered currency account which holds
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the market
 * @param currencyMint - Registered currency's mint
 * @param royalty - Royalty rate for exchanges denominated in this currency (Royalty of '1_000_000' == 100%)
 * @param saCurrencyVault - Token account to which royalties for this currency are to be sent
 * @returns InstructionReturn
 */
export function registerCurrency(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  currencyMint: PublicKey,
  royalty: number,
  saCurrencyVault: PublicKey,
): InstructionReturn {
  return async () => ({
    instruction: await program.methods
      .registerCurrency(new BN(royalty))
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        systemProgram: SystemProgram.programId,
        registeredCurrency: RegisteredCurrency.findAddress(
          program,
          currencyMint,
        )[0],
        currencyMint,
        saCurrencyVault,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}
