// @staratlas/galactic-marketplace/src/instructions/removeFeeExemption.ts
import { PublicKey } from '@solana/web3.js';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram, MarketVars } from '..';
import { FeeReduction } from '../state/feeReduction';

/**
 *
 * @param program - Marketplace program
 * @param updateAuthorityAccount - Public key of registered update authority
 * @param feeExemptTarget - Target account to exempt from fees
 * @returns InstructionReturn
 */
export function removeFeeExemption(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  feeExemptTarget: PublicKey,
): InstructionReturn {
  const feeExemptAccount = FeeReduction.findAddress(
    feeExemptTarget,
    program,
  )[0];
  const marketVarsAccount = MarketVars.findAddress(program)[0];
  return async (funder) => [
    {
      instruction: await program.methods
        .removeFeeExemption()
        .accountsStrict({
          updateAuthorityMaster: updateAuthorityAccount.publicKey(),
          funder: funder.publicKey(),
          feeExemptAccount,
          marketVarsAccount,
        })
        .instruction(),
      signers: [updateAuthorityAccount],
    },
  ];
}
