// @staratlas/galactic-marketplace/src/instructions/royaltyTiers.ts
import { PublicKey } from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars, RegisteredCurrency } from '../state';

/**
 * Add a new royalty tier
 * This associates a quantity of staked tokens with a marketplace
 * royalty fee discount if it does not already exist
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the market
 * @param registeredCurrency - a valid registered currency
 * @param stakeAmount - the staked amount in the token's base units
 * @param discount - the discount amount (Discount of '1_000_000' == 100%)
 * @returns InstructionReturn
 */
export function addRoyaltyTier(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  registeredCurrency:
    | { registeredCurrency: RegisteredCurrency }
    | { key: PublicKey },
  stakeAmount: number,
  discount: number,
): InstructionReturn {
  const registeredCurrencyAddress =
    'key' in registeredCurrency
      ? registeredCurrency.key
      : registeredCurrency.registeredCurrency.key;
  return async () => ({
    instruction: await program.methods
      .addRoyaltyTier(new BN(stakeAmount), new BN(discount))
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        registeredCurrency: registeredCurrencyAddress,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}

/**
 * Delete an existing royalty tier
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the market
 * @param registeredCurrency - a valid registered currency
 * @param stakeAmount - the staked amount in the token's base units
 * @returns InstructionReturn
 */
export function deleteRoyaltyTier(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  registeredCurrency:
    | { registeredCurrency: RegisteredCurrency }
    | { key: PublicKey },
  stakeAmount: number,
): InstructionReturn {
  const registeredCurrencyAddress =
    'key' in registeredCurrency
      ? registeredCurrency.key
      : registeredCurrency.registeredCurrency.key;
  return async () => ({
    instruction: await program.methods
      .deleteRoyaltyTier(new BN(stakeAmount))
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        registeredCurrency: registeredCurrencyAddress,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}

/**
 * Update the discount rate associated with a given royalty tier if it exists
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the market
 * @param registeredCurrency - a valid registered currency
 * @param stakeAmount - the staked amount in the token's base units
 * @param discount - the new discount amount (Discount of '1_000_000' == 100%)
 * @returns InstructionReturn
 */
export function updateRoyaltyTier(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  registeredCurrency:
    | { registeredCurrency: RegisteredCurrency }
    | { key: PublicKey },
  stakeAmount: number,
  discount: number,
): InstructionReturn {
  const registeredCurrencyAddress =
    'key' in registeredCurrency
      ? registeredCurrency.key
      : registeredCurrency.registeredCurrency.key;
  return async () => ({
    instruction: await program.methods
      .updateRoyaltyTier(new BN(stakeAmount), new BN(discount))
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        registeredCurrency: registeredCurrencyAddress,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}
