// @staratlas/galactic-marketplace/src/instructions/updateCurrency.ts
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars, RegisteredCurrency } from '../state';

/**
 * Used to update the royalty rate for a registered currency
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the market
 * @param currencyMint - Registered currency's mint
 * @param royalty - Royalty rate for exchanges denominated in this currency (Royalty of '1_000_000' == 100%)
 * @returns InstructionReturn
 */
export function updateRegisteredCurrencyRoyalty(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  currencyMint: PublicKey,
  royalty: number,
): InstructionReturn {
  return async () => ({
    instruction: await program.methods
      .updateCurrencyRoyalty(new BN(royalty))
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        registeredCurrency: RegisteredCurrency.findAddress(
          program,
          currencyMint,
        )[0],
        currencyMint,
        systemProgram: SystemProgram.programId,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}

/**
 * Updates the Currency Vault associated with a RegisteredCurrency
 * @param program - GalacticMarketplace Program
 * @param updateAuthorityAccount - the authority of the market
 * @param currencyMint - Registered currency's mint
 * @param saCurrencyVault - Token account to which royalties for this currency are to be sent
 * @returns InstructionReturn
 */
export function updateRegisteredCurrencyVault(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  currencyMint: PublicKey,
  saCurrencyVault: PublicKey,
): InstructionReturn {
  return async () => ({
    instruction: await program.methods
      .updateCurrencyVault()
      .accountsStrict({
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        marketVarsAccount: MarketVars.findAddress(program)[0],
        registeredCurrency: RegisteredCurrency.findAddress(
          program,
          currencyMint,
        )[0],
        currencyMint,
        systemProgram: SystemProgram.programId,
        saCurrencyVault,
      })
      .instruction(),
    signers: [updateAuthorityAccount],
  });
}
