// @staratlas/galactic-marketplace/src/instructions/updateAtlasRate.ts
import { SystemProgram } from '@solana/web3.js';
import { AsyncSigner, InstructionReturn } from '@staratlas/data-source';
import BN from 'bn.js';
import { GalacticMarketplaceIDLProgram } from '../constants';
import { MarketVars } from '../state';
import { AtlasRate } from '../state/atlasRateAccount';

/**
 * `atlas_rate` * Atlas sub-tokens = USDC sub-tokens
 * `atlas_rate` = `1`/(`100` * Atlas price in USD)
 * @param program - The GalacticMarketplace program
 * @param updateAuthorityAccount - The authority of the market
 * @param rate - The rate of Atlas to USDC
 * @returns InstructionReturn
 */
export function updateAtlasRate(
  program: GalacticMarketplaceIDLProgram,
  updateAuthorityAccount: AsyncSigner,
  rate: number,
): InstructionReturn {
  return async (funder) => ({
    instruction: await program.methods
      .updateAtlasRate(new BN(Math.floor(rate * 1_000_000_000)))
      .accountsStrict({
        funder: funder.publicKey(),
        updateAuthorityAccount: updateAuthorityAccount.publicKey(),
        atlasRate: AtlasRate.findAddress(program)[0],
        marketVarsAccount: MarketVars.findAddress(program)[0],
        systemProgram: SystemProgram.programId,
      })
      .instruction(),
    signers: [updateAuthorityAccount, funder],
  });
}
