// @staratlas/galactic-marketplace/dist/src/util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findExtraAccountMetaListAddress = exports.createOrderAccountInstruction = exports.getTokenAccount = exports.getOrderSide = void 0;
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const anchor_1 = require("@staratlas/anchor");
const data_source_1 = require("@staratlas/data-source");
const state_1 = require("./state");
const getOrderSide = (orderAccount) => {
    if (JSON.stringify(orderAccount.data.orderSide) === JSON.stringify({ buy: {} })) {
        return 'BuySide';
    }
    return 'SellSide';
};
exports.getOrderSide = getOrderSide;
/**
 * Get or create a token account
 * @param connection - Solana connection object
 * @param wallet - the owner of the token account
 * @param mint - the mint
 * @param newAccountFunder - the payer of the rent for new account if a new account is to be created
 * @param amountNeededHeuristic - amount needed heuristic. Used to select the minimum acceptable balance in case the wallet has more than one token account
 * @param programId - the token program id
 * @returns the token account address
 */
function getTokenAccount(connection_1, wallet_1, mint_1) {
    return __awaiter(this, arguments, void 0, function* (connection, wallet, mint, newAccountFunder = wallet, amountNeededHeuristic, programId = spl_token_1.TOKEN_PROGRAM_ID) {
        // Get all parsed token accounts for a given owner and mint
        const { value: tokenAccounts } = yield connection.getParsedTokenAccountsByOwner(wallet, { mint, programId });
        let tokenAccount = null;
        if (amountNeededHeuristic === undefined) {
            let maxTokens = null;
            for (const account of tokenAccounts) {
                if (account.account.data.parsed.type === 'account') {
                    const parsedAccount = account.account.data.parsed;
                    const tokenAmount = new anchor_1.BN(parsedAccount.info.tokenAmount.amount);
                    if (maxTokens === null || tokenAmount.gt(maxTokens)) {
                        tokenAccount = account.pubkey;
                        maxTokens = tokenAmount;
                    }
                }
            }
        }
        else {
            let minTokens = null;
            for (const account of tokenAccounts) {
                if (account.account.data.parsed.type === 'account') {
                    const parsedAccount = account.account.data.parsed;
                    const tokenAmount = new anchor_1.BN(parsedAccount.info.tokenAmount.amount);
                    if (tokenAmount.gt(new anchor_1.BN(amountNeededHeuristic)) &&
                        (minTokens === null || tokenAmount.lt(minTokens))) {
                        tokenAccount = account.pubkey;
                        minTokens = tokenAmount;
                    }
                }
            }
        }
        if (tokenAccount === null) {
            const associatedTokenAddress = (0, spl_token_1.getAssociatedTokenAddressSync)(mint, wallet, true, programId);
            const { value: account } = yield connection.getParsedAccountInfo(associatedTokenAddress, 'confirmed');
            if (account === null || account.owner.equals(web3_js_1.SystemProgram.programId)) {
                return {
                    tokenAccount: associatedTokenAddress,
                    createInstruction: () => Promise.resolve({
                        instruction: (0, spl_token_1.createAssociatedTokenAccountInstruction)(newAccountFunder, associatedTokenAddress, wallet, mint, programId, spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID),
                        signers: [],
                    }),
                };
            }
            if ('program' in account.data &&
                account.data.parsed.type === 'account' &&
                new web3_js_1.PublicKey(account.data.parsed.info.owner).equals(wallet)) {
                return {
                    tokenAccount: associatedTokenAddress,
                };
            }
            else {
                const newTokenAccount = web3_js_1.Keypair.generate();
                return {
                    tokenAccount: newTokenAccount,
                    createInstruction: () => Promise.resolve({
                        instruction: (0, spl_token_1.createInitializeAccountInstruction)(newTokenAccount.publicKey, mint, wallet, programId),
                        signers: [(0, data_source_1.keypairToAsyncSigner)(newTokenAccount)],
                    }),
                };
            }
        }
        else {
            return {
                tokenAccount,
            };
        }
    });
}
exports.getTokenAccount = getTokenAccount;
/**
 * Create a new empty order account
 * @param connection - Solana connection object
 * @param program - GalacticMarketplace Program
 * @param orderAccount - the order account address Keypair
 * @returns a promise of InstructionReturn
 */
function createOrderAccountInstruction(connection, program, orderAccount) {
    return Promise.resolve((funder) => __awaiter(this, void 0, void 0, function* () {
        return ({
            instruction: web3_js_1.SystemProgram.createAccount({
                fromPubkey: funder.publicKey(),
                newAccountPubkey: orderAccount.publicKey,
                space: state_1.Order.MIN_DATA_SIZE,
                lamports: yield connection.getMinimumBalanceForRentExemption(state_1.Order.MIN_DATA_SIZE),
                programId: program.programId,
            }),
            signers: [(0, data_source_1.keypairToAsyncSigner)(orderAccount)],
        });
    }));
}
exports.createOrderAccountInstruction = createOrderAccountInstruction;
/**
 * Find the transfer hook account meta list address
 * @param transferHookProgram - the transfer hook program
 * @param mint - the cargo mint
 * @returns The PDA and bump respectively
 */
function findExtraAccountMetaListAddress(transferHookProgram, mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([Buffer.from('extra-account-metas'), mint.toBuffer()], transferHookProgram);
}
exports.findExtraAccountMetaListAddress = findExtraAccountMetaListAddress;
