// @staratlas/galactic-marketplace/src/constants.ts
import { PublicKey } from '@solana/web3.js';
import { Coder, Program, Provider } from '@staratlas/anchor';
import {
  AnchorTypes,
  ExtractArrayType,
  ListenProgram,
  ListenProgramStatic,
  ProgramMethods,
  staticImplements,
} from '@staratlas/data-source';
import {
  IDL as GALACTIC_MARKETPLACE_IDL,
  Marketplace as GalacticMarketplaceIDL,
} from './idl/marketplace';
import {
  FeeReduction,
  MarketVars,
  OpenOrdersCounter,
  Order,
  RegisteredCurrency,
} from './state';

export { IDL as GALACTIC_MARKETPLACE_IDL } from './idl/marketplace';
export type { Marketplace as GalacticMarketplaceIDL } from './idl/marketplace';

export type GalacticMarketplaceAccountsArray = ExtractArrayType<
  GalacticMarketplaceIDL['accounts']
>['name'];
export type GalacticMarketplaceTypesArray = ExtractArrayType<
  GalacticMarketplaceIDL['types']
>['name'];

export type GalacticMarketplaceIDLProgram =
  ProgramMethods<GalacticMarketplaceIDL>;
export type GalacticMarketplaceCoder = Coder<
  GalacticMarketplaceAccountsArray,
  GalacticMarketplaceTypesArray
>;

export type GalacticMarketplaceTypes = AnchorTypes<GalacticMarketplaceIDL>;

// Accounts
export type GalacticMarketplaceIDLAccounts =
  GalacticMarketplaceTypes['Accounts'];
export type MarketVarsAccount = GalacticMarketplaceIDLAccounts['marketVars'];
export type OpenOrdersCounterAccount =
  GalacticMarketplaceIDLAccounts['openOrdersCounter'];
export type OrderAccount = GalacticMarketplaceIDLAccounts['orderAccount'];
export type RegisteredCurrencyAccount =
  GalacticMarketplaceIDLAccounts['registeredCurrency'];
export type FeeReductionAccount =
  GalacticMarketplaceIDLAccounts['feeReduction'];
export type AtlasRateAccount =
  GalacticMarketplaceIDLAccounts['atlasRateAccount'];

// Instructions
export type GalacticMarketplaceInstructions =
  GalacticMarketplaceTypes['Instructions'];
export type AddRoyaltyTier = GalacticMarketplaceInstructions['addRoyaltyTier'];
export type DeleteRoyaltyTier =
  GalacticMarketplaceInstructions['deleteRoyaltyTier'];
export type DeregisterCurrency =
  GalacticMarketplaceInstructions['deregisterCurrency'];
export type InitializeMarketplace =
  GalacticMarketplaceInstructions['initializeMarketplace'];
export type RegisterCurrency =
  GalacticMarketplaceInstructions['registerCurrency'];
export type UpdateCurrencyVault =
  GalacticMarketplaceInstructions['updateCurrencyVault'];
export type UpdateCurrencyRoyalty =
  GalacticMarketplaceInstructions['updateCurrencyRoyalty'];
export type UpdateRoyaltyTier =
  GalacticMarketplaceInstructions['updateRoyaltyTier'];
export type ProcessInitializeBuy =
  GalacticMarketplaceInstructions['processInitializeBuy'];
export type ProcessInitializeSell =
  GalacticMarketplaceInstructions['processInitializeSell'];
export type ProcessExchange =
  GalacticMarketplaceInstructions['processExchange'];
export type ProcessCancel = GalacticMarketplaceInstructions['processCancel'];
export type InitializeOpenOrderCounter =
  GalacticMarketplaceInstructions['initializeOpenOrdersCounter'];
export type AddFeeExemption =
  GalacticMarketplaceInstructions['addFeeExemption'];
export type RemoveFeeExemption =
  GalacticMarketplaceInstructions['removeFeeExemption'];

// Types
export type RoyaltyTier = GalacticMarketplaceTypes['Defined']['RoyaltyTier'];
export type OrderSide = GalacticMarketplaceTypes['Defined']['OrderSide'];

export type GalacticMarketplaceAccounts = {
  marketVars: MarketVars;
  openOrdersCounter: OpenOrdersCounter;
  orderAccount: Order;
  registeredCurrency: RegisteredCurrency;
  feeReduction: FeeReduction;
};

@staticImplements<
  ListenProgramStatic<
    GalacticMarketplaceProgram,
    GalacticMarketplaceAccounts,
    GalacticMarketplaceIDL
  >
>()
export class GalacticMarketplaceProgram extends ListenProgram<
  GalacticMarketplaceAccounts,
  GalacticMarketplaceIDL
> {
  constructor(program: GalacticMarketplaceIDLProgram) {
    super(program, {
      marketVars: MarketVars,
      openOrdersCounter: OpenOrdersCounter,
      orderAccount: Order,
      registeredCurrency: RegisteredCurrency,
      feeReduction: FeeReduction,
    });
  }

  static buildProgram(
    programId: PublicKey,
    provider?: Provider,
    coder?: Coder,
  ): GalacticMarketplaceIDLProgram {
    return new Program(GALACTIC_MARKETPLACE_IDL, programId, provider, coder);
  }
}
