// @staratlas/galactic-marketplace/src/util.ts
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  createAssociatedTokenAccountInstruction,
  createInitializeAccountInstruction,
  getAssociatedTokenAddressSync,
  TOKEN_PROGRAM_ID,
} from '@solana/spl-token';
import { Connection, Keypair, PublicKey, SystemProgram } from '@solana/web3.js';
import { BN } from '@staratlas/anchor';
import {
  InstructionReturn,
  keypairToAsyncSigner,
} from '@staratlas/data-source';
import { GalacticMarketplaceIDLProgram } from './constants';
import { Order } from './state';

type TokenAccount = {
  type: 'account';
  info: {
    isNative: boolean;
    mint: string;
    owner: string;
    state: 'initialized';
    tokenAmount: {
      amount: string;
      decimals: number;
      uiAmount: number;
      uiAmountString: string;
    };
  };
};

export const getOrderSide = (orderAccount: Order) => {
  if (
    JSON.stringify(orderAccount.data.orderSide) === JSON.stringify({ buy: {} })
  ) {
    return 'BuySide';
  }
  return 'SellSide';
};

/**
 * Get or create a token account
 * @param connection - Solana connection object
 * @param wallet - the owner of the token account
 * @param mint - the mint
 * @param newAccountFunder - the payer of the rent for new account if a new account is to be created
 * @param amountNeededHeuristic - amount needed heuristic. Used to select the minimum acceptable balance in case the wallet has more than one token account
 * @param programId - the token program id
 * @returns the token account address
 */
export async function getTokenAccount(
  connection: Connection,
  wallet: PublicKey,
  mint: PublicKey,
  newAccountFunder: PublicKey = wallet,
  amountNeededHeuristic?: number,
  programId = TOKEN_PROGRAM_ID,
): Promise<
  | {
      tokenAccount: PublicKey;
    }
  | {
      tokenAccount: Keypair | PublicKey;
      createInstruction: InstructionReturn;
    }
> {
  // Get all parsed token accounts for a given owner and mint
  const { value: tokenAccounts } =
    await connection.getParsedTokenAccountsByOwner(wallet, { mint, programId });
  let tokenAccount: PublicKey | null = null;
  if (amountNeededHeuristic === undefined) {
    let maxTokens: BN | null = null;
    for (const account of tokenAccounts) {
      if (account.account.data.parsed.type === 'account') {
        const parsedAccount = account.account.data.parsed as TokenAccount;
        const tokenAmount = new BN(parsedAccount.info.tokenAmount.amount);
        if (maxTokens === null || tokenAmount.gt(maxTokens)) {
          tokenAccount = account.pubkey;
          maxTokens = tokenAmount;
        }
      }
    }
  } else {
    let minTokens: BN | null = null;
    for (const account of tokenAccounts) {
      if (account.account.data.parsed.type === 'account') {
        const parsedAccount = account.account.data.parsed as TokenAccount;
        const tokenAmount = new BN(parsedAccount.info.tokenAmount.amount);
        if (
          tokenAmount.gt(new BN(amountNeededHeuristic)) &&
          (minTokens === null || tokenAmount.lt(minTokens))
        ) {
          tokenAccount = account.pubkey;
          minTokens = tokenAmount;
        }
      }
    }
  }

  if (tokenAccount === null) {
    const associatedTokenAddress = getAssociatedTokenAddressSync(
      mint,
      wallet,
      true,
      programId,
    );
    const { value: account } = await connection.getParsedAccountInfo(
      associatedTokenAddress,
      'confirmed',
    );
    if (account === null || account.owner.equals(SystemProgram.programId)) {
      return {
        tokenAccount: associatedTokenAddress,
        createInstruction: () =>
          Promise.resolve({
            instruction: createAssociatedTokenAccountInstruction(
              newAccountFunder,
              associatedTokenAddress,
              wallet,
              mint,
              programId,
              ASSOCIATED_TOKEN_PROGRAM_ID,
            ),
            signers: [],
          }),
      };
    }

    if (
      'program' in account.data &&
      account.data.parsed.type === 'account' &&
      new PublicKey((account.data.parsed as TokenAccount).info.owner).equals(
        wallet,
      )
    ) {
      return {
        tokenAccount: associatedTokenAddress,
      };
    } else {
      const newTokenAccount = Keypair.generate();
      return {
        tokenAccount: newTokenAccount,
        createInstruction: () =>
          Promise.resolve({
            instruction: createInitializeAccountInstruction(
              newTokenAccount.publicKey,
              mint,
              wallet,
              programId,
            ),
            signers: [keypairToAsyncSigner(newTokenAccount)],
          }),
      };
    }
  } else {
    return {
      tokenAccount,
    };
  }
}

/**
 * Create a new empty order account
 * @param connection - Solana connection object
 * @param program - GalacticMarketplace Program
 * @param orderAccount - the order account address Keypair
 * @returns a promise of InstructionReturn
 */
export function createOrderAccountInstruction(
  connection: Connection,
  program: GalacticMarketplaceIDLProgram,
  orderAccount: Keypair,
): Promise<InstructionReturn> {
  return Promise.resolve(async (funder) => ({
    instruction: SystemProgram.createAccount({
      fromPubkey: funder.publicKey(),
      newAccountPubkey: orderAccount.publicKey,
      space: Order.MIN_DATA_SIZE,
      lamports: await connection.getMinimumBalanceForRentExemption(
        Order.MIN_DATA_SIZE,
      ),
      programId: program.programId,
    }),
    signers: [keypairToAsyncSigner(orderAccount)],
  }));
}

/**
 * Find the transfer hook account meta list address
 * @param transferHookProgram - the transfer hook program
 * @param mint - the cargo mint
 * @returns The PDA and bump respectively
 */
export function findExtraAccountMetaListAddress(
  transferHookProgram: PublicKey,
  mint: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('extra-account-metas'), mint.toBuffer()],
    transferHookProgram,
  );
}
