// @staratlas/galactic-marketplace/dist/src/state/registeredCurrency.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var RegisteredCurrency_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisteredCurrency = exports.VEC_ROYALTY_TIER_SIZE = exports.ROYALTY_TIER_SIZE = exports.registeredCurrencyDataEquals = void 0;
const web3_js_1 = require("@solana/web3.js");
const data_source_1 = require("@staratlas/data-source");
/**
 * Check if two instances of `RegisteredCurrencyAccount` are equal
 * @param registeredCurrencyData1 - the first instance
 * @param registeredCurrencyData2 - the second instance
 * @returns boolean
 */
function registeredCurrencyDataEquals(registeredCurrencyData1, registeredCurrencyData2) {
    return (registeredCurrencyData1.tokenMint === registeredCurrencyData2.tokenMint &&
        registeredCurrencyData1.saCurrencyVault ===
            registeredCurrencyData2.saCurrencyVault &&
        registeredCurrencyData1.royalty === registeredCurrencyData2.royalty &&
        registeredCurrencyData1.royaltyTiers ===
            registeredCurrencyData2.royaltyTiers &&
        registeredCurrencyData1.bump === registeredCurrencyData2.bump);
}
exports.registeredCurrencyDataEquals = registeredCurrencyDataEquals;
exports.ROYALTY_TIER_SIZE = 8 + // stake amount
    8; // discount
exports.VEC_ROYALTY_TIER_SIZE = 1 + // 1 for optional royalty tiers
    4 + // 4 for vector size
    100 * // Number of royalty tiers
        exports.ROYALTY_TIER_SIZE;
let RegisteredCurrency = RegisteredCurrency_1 = class RegisteredCurrency {
    constructor(_data, _key) {
        this._data = _data;
        this._key = _key;
    }
    get data() {
        return this._data;
    }
    get key() {
        return this._key;
    }
    /**
     * Find the `RegisteredCurrency` account address
     * @param program - GalacticMarketplace Program
     * @param currencyMint - the current mint
     * @returns PDA and bump
     */
    static findAddress(program, currencyMint) {
        return web3_js_1.PublicKey.findProgramAddressSync([Buffer.from('registered-currency'), currencyMint.toBuffer()], program.programId);
    }
    static decodeData(account, program) {
        return (0, data_source_1.decodeAccount)(account, program, RegisteredCurrency_1);
    }
};
exports.RegisteredCurrency = RegisteredCurrency;
RegisteredCurrency.ACCOUNT_NAME = 'registeredCurrency';
RegisteredCurrency.MIN_DATA_SIZE = 8 + // discriminator
    32 + // token mint
    32 + // sa currency vault
    8 + // royalty
    exports.VEC_ROYALTY_TIER_SIZE + // royalty tiers
    1; // bump
exports.RegisteredCurrency = RegisteredCurrency = RegisteredCurrency_1 = __decorate([
    (0, data_source_1.staticImplements)(),
    __metadata("design:paramtypes", [Object, web3_js_1.PublicKey])
], RegisteredCurrency);
