// @staratlas/galactic-marketplace/src/state/atlasRateAccount.ts
import { KeyedAccountInfo, PublicKey } from '@solana/web3.js';
import {
  Account,
  AccountStatic,
  DecodedAccountData,
  decodeAccount,
  staticImplements,
} from '@staratlas/data-source';
import {
  AtlasRateAccount,
  GalacticMarketplaceIDL,
  GalacticMarketplaceIDLProgram,
} from '../constants';
export const ATLAS_RATE_ACCOUNT_MIN_DATA_SIZE =
  8 + // discriminator
  8; // atlas_rate

@staticImplements<AccountStatic<AtlasRate, GalacticMarketplaceIDL>>()
export class AtlasRate implements Account {
  static readonly ACCOUNT_NAME: NonNullable<
    GalacticMarketplaceIDL['accounts']
  >[number]['name'] = 'atlasRateAccount';
  static readonly MIN_DATA_SIZE: number = ATLAS_RATE_ACCOUNT_MIN_DATA_SIZE;

  constructor(private _data: AtlasRateAccount, private _key: PublicKey) {}

  get data(): Readonly<AtlasRateAccount> {
    return this._data;
  }

  get key(): Readonly<PublicKey> {
    return this._key;
  }

  /**
   * Find the AtlasRate account address
   * @param program - Marketplace program
   * @returns PDA for Atlas Rate Account
   */
  static findAddress(
    program: GalacticMarketplaceIDLProgram,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('atlas_rate')],
      program.programId,
    );
  }
  static decodeData(
    account: KeyedAccountInfo,
    program: GalacticMarketplaceIDLProgram,
  ): DecodedAccountData<AtlasRate> {
    return decodeAccount(account, program, AtlasRate);
  }
}
