// @staratlas/galactic-marketplace/src/state/feeReduction.ts
import { KeyedAccountInfo, PublicKey } from '@solana/web3.js';
import {
  Account,
  AccountStatic,
  DecodedAccountData,
  decodeAccount,
  staticImplements,
} from '@staratlas/data-source';
import {
  FeeReductionAccount,
  GalacticMarketplaceIDL,
  GalacticMarketplaceIDLProgram,
} from '../constants';
export const FEE_REDUCTION_MIN_DATA_SIZE =
  8 + // discriminator
  32 + // account key
  1 + // bump
  8; // discount

@staticImplements<AccountStatic<FeeReduction, GalacticMarketplaceIDL>>()
export class FeeReduction implements Account {
  static readonly ACCOUNT_NAME: NonNullable<
    GalacticMarketplaceIDL['accounts']
  >[number]['name'] = 'feeReduction';
  static readonly MIN_DATA_SIZE: number = FEE_REDUCTION_MIN_DATA_SIZE;

  constructor(private _data: FeeReductionAccount, private _key: PublicKey) {}

  get data(): Readonly<FeeReductionAccount> {
    return this._data;
  }

  get key(): Readonly<PublicKey> {
    return this._key;
  }

  /**
   * Find the FeeReduction account address
   * @param targetAccount - Account key
   * @param program - Marketplace program
   * @returns PDA for Fee Reduction account
   */
  static findAddress(
    targetAccount: PublicKey,
    program: GalacticMarketplaceIDLProgram,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('fee_exempt'), targetAccount.toBuffer()],
      program.programId,
    );
  }
  static decodeData(
    account: KeyedAccountInfo,
    program: GalacticMarketplaceIDLProgram,
  ): DecodedAccountData<FeeReduction> {
    return decodeAccount(account, program, FeeReduction);
  }
}
