// @staratlas/galactic-marketplace/src/state/marketVars.ts
import { KeyedAccountInfo, PublicKey } from '@solana/web3.js';
import {
  Account,
  AccountStatic,
  DecodedAccountData,
  decodeAccount,
  staticImplements,
} from '@staratlas/data-source';
import { GalacticMarketplaceIDL } from '..';
import { GalacticMarketplaceIDLProgram, MarketVarsAccount } from '../constants';

/**
 * Check if two instances of `MarketVarsAccount` are equal
 * @param marketVarsData1 - the first instance
 * @param marketVarsData2 - the second instance
 * @returns boolean
 */
export function marketVarsDataEquals(
  marketVarsData1: MarketVarsAccount,
  marketVarsData2: MarketVarsAccount,
): boolean {
  return (
    marketVarsData1.updateAuthorityMaster ===
      marketVarsData2.updateAuthorityMaster &&
    marketVarsData1.bump === marketVarsData2.bump
  );
}

@staticImplements<AccountStatic<MarketVars, GalacticMarketplaceIDL>>()
export class MarketVars implements Account {
  static readonly ACCOUNT_NAME: NonNullable<
    GalacticMarketplaceIDL['accounts']
  >[number]['name'] = 'marketVars';
  static readonly MIN_DATA_SIZE =
    8 + // discriminator
    32 + // update authority master
    1; // bump

  constructor(private _data: MarketVarsAccount, private _key: PublicKey) {}

  get data(): Readonly<MarketVarsAccount> {
    return this._data;
  }

  get key(): PublicKey {
    return this._key;
  }

  /**
   * Find the `MarketVars` account address
   * @param program - GalacticMarketplace Program
   * @returns PDA and bump
   */
  static findAddress(
    program: GalacticMarketplaceIDLProgram,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('market-vars')],
      program.programId,
    );
  }

  static decodeData(
    account: KeyedAccountInfo,
    program: GalacticMarketplaceIDLProgram,
  ): DecodedAccountData<MarketVars> {
    return decodeAccount(account, program, MarketVars);
  }
}
