// @staratlas/galactic-marketplace/src/state/openOrdersCounter.ts
import { KeyedAccountInfo, PublicKey } from '@solana/web3.js';
import {
  Account,
  AccountStatic,
  DecodedAccountData,
  decodeAccount,
  staticImplements,
} from '@staratlas/data-source';
import { GalacticMarketplaceIDL } from '..';
import {
  GalacticMarketplaceIDLProgram,
  OpenOrdersCounterAccount,
} from '../constants';

/**
 * Check if two instances of `OpenOrdersCounterAccount` are equal
 * @param openOrdersCounterData1 - the first instance
 * @param openOrdersCounterData2 - the second instance
 * @returns boolean
 */
export function openOrdersCounterDataEquals(
  openOrdersCounterData1: OpenOrdersCounterAccount,
  openOrdersCounterData2: OpenOrdersCounterAccount,
): boolean {
  return (
    openOrdersCounterData1.openOrderCount ===
      openOrdersCounterData2.openOrderCount &&
    openOrdersCounterData1.bump === openOrdersCounterData2.bump
  );
}

@staticImplements<AccountStatic<OpenOrdersCounter, GalacticMarketplaceIDL>>()
export class OpenOrdersCounter implements Account {
  static readonly ACCOUNT_NAME: NonNullable<
    GalacticMarketplaceIDL['accounts']
  >[number]['name'] = 'openOrdersCounter';
  static readonly MIN_DATA_SIZE =
    8 + // discriminator
    8 + // counter
    1; // bump

  constructor(
    private _data: OpenOrdersCounterAccount,
    private _key: PublicKey,
  ) {}

  get data(): Readonly<OpenOrdersCounterAccount> {
    return this._data;
  }

  get key(): PublicKey {
    return this._key;
  }

  /**
   * Find the `OpenOrdersCounter` account address
   * @param program - GalacticMarketplace Program
   * @param playerPubkey - the player address
   * @param depositMint - the deposit mint address
   * @returns PDA and bump
   */
  static findAddress(
    program: GalacticMarketplaceIDLProgram,
    playerPubkey: PublicKey,
    depositMint: PublicKey,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('open-orders-counter'),
        playerPubkey.toBuffer(),
        depositMint.toBuffer(),
      ],
      program.programId,
    );
  }

  static decodeData(
    account: KeyedAccountInfo,
    program: GalacticMarketplaceIDLProgram,
  ): DecodedAccountData<OpenOrdersCounter> {
    return decodeAccount(account, program, OpenOrdersCounter);
  }
}
