// @staratlas/galactic-marketplace/src/state/orderAccount.ts
import { KeyedAccountInfo, PublicKey } from '@solana/web3.js';
import {
  Account,
  AccountStatic,
  DecodedAccountData,
  decodeAccount,
  staticImplements,
} from '@staratlas/data-source';
import { GalacticMarketplaceIDL } from '..';
import { GalacticMarketplaceIDLProgram, OrderAccount } from '../constants';

/**
 * Check if two instances of `OrderAccount` are equal
 * @param orderData1 - the first instance
 * @param orderData2 - the second instance
 * @returns boolean
 */
export function orderDataEquals(
  orderData1: OrderAccount,
  orderData2: OrderAccount,
): boolean {
  return (
    orderData1.orderInitializerPubkey === orderData2.orderInitializerPubkey &&
    orderData1.currencyMint === orderData2.currencyMint &&
    orderData1.assetMint === orderData2.assetMint &&
    orderData1.initializerCurrencyTokenAccount ===
      orderData2.initializerCurrencyTokenAccount &&
    orderData1.initializerAssetTokenAccount ===
      orderData2.initializerAssetTokenAccount &&
    orderData1.orderSide === orderData2.orderSide &&
    orderData1.price === orderData2.price &&
    orderData1.orderOriginationQty === orderData2.orderOriginationQty &&
    orderData1.orderRemainingQty === orderData2.orderRemainingQty &&
    orderData1.createdAtTimestamp === orderData2.createdAtTimestamp
  );
}

@staticImplements<AccountStatic<Order, GalacticMarketplaceIDL>>()
export class Order implements Account {
  static readonly ACCOUNT_NAME: NonNullable<
    GalacticMarketplaceIDL['accounts']
  >[number]['name'] = 'orderAccount';
  static readonly MIN_DATA_SIZE =
    8 + // discriminator
    32 + // order initializer pubkey
    32 + // currency mint
    32 + // asset mint
    32 + // initializer currency token account
    32 + // initializer asset token account
    1 + // order side
    8 + // price
    8 + // order origination qty
    8 + // order remaining qty
    8; // created at timestamp

  constructor(private _data: OrderAccount, private _key: PublicKey) {}

  /**
   * Find the `OrderVault` account address
   * @param program - GalacticMarketplace Program
   * @param orderInitializer - the order initializer
   * @param depositMint - the deposit mint
   * @returns PDA and bump
   */
  static getOrderVaultAddress(
    program: GalacticMarketplaceIDLProgram,
    orderInitializer: PublicKey,
    depositMint: PublicKey,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('order-vault-account'),
        orderInitializer.toBuffer(),
        depositMint.toBuffer(),
      ],
      program.programId,
    );
  }

  /**
   * Find the `OrderVaultAuthority` account address
   * @param program - GalacticMarketplace Program
   * @param orderInitializer - the order initializer
   * @returns PDA and bump
   */
  static getOrderVaultAuthorityAddress(
    program: GalacticMarketplaceIDLProgram,
    orderInitializer: PublicKey,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('order-vault-auth'), orderInitializer.toBuffer()],
      program.programId,
    );
  }

  get data(): Readonly<OrderAccount> {
    return this._data;
  }

  get key(): PublicKey {
    return this._key;
  }

  static decodeData(
    account: KeyedAccountInfo,
    program: GalacticMarketplaceIDLProgram,
  ): DecodedAccountData<Order> {
    return decodeAccount(account, program, Order);
  }
}
