// @staratlas/galactic-marketplace/src/state/registeredCurrency.ts
import { KeyedAccountInfo, PublicKey } from '@solana/web3.js';
import {
  Account,
  AccountStatic,
  DecodedAccountData,
  decodeAccount,
  staticImplements,
} from '@staratlas/data-source';
import { GalacticMarketplaceIDL } from '..';
import {
  GalacticMarketplaceIDLProgram,
  RegisteredCurrencyAccount,
} from '../constants';

/**
 * Check if two instances of `RegisteredCurrencyAccount` are equal
 * @param registeredCurrencyData1 - the first instance
 * @param registeredCurrencyData2 - the second instance
 * @returns boolean
 */
export function registeredCurrencyDataEquals(
  registeredCurrencyData1: RegisteredCurrencyAccount,
  registeredCurrencyData2: RegisteredCurrencyAccount,
): boolean {
  return (
    registeredCurrencyData1.tokenMint === registeredCurrencyData2.tokenMint &&
    registeredCurrencyData1.saCurrencyVault ===
      registeredCurrencyData2.saCurrencyVault &&
    registeredCurrencyData1.royalty === registeredCurrencyData2.royalty &&
    registeredCurrencyData1.royaltyTiers ===
      registeredCurrencyData2.royaltyTiers &&
    registeredCurrencyData1.bump === registeredCurrencyData2.bump
  );
}

export const ROYALTY_TIER_SIZE =
  8 + // stake amount
  8; // discount

export const VEC_ROYALTY_TIER_SIZE =
  1 + // 1 for optional royalty tiers
  4 + // 4 for vector size
  100 * // Number of royalty tiers
    ROYALTY_TIER_SIZE;

@staticImplements<AccountStatic<RegisteredCurrency, GalacticMarketplaceIDL>>()
export class RegisteredCurrency implements Account {
  static readonly ACCOUNT_NAME: NonNullable<
    GalacticMarketplaceIDL['accounts']
  >[number]['name'] = 'registeredCurrency';
  static readonly MIN_DATA_SIZE =
    8 + // discriminator
    32 + // token mint
    32 + // sa currency vault
    8 + // royalty
    VEC_ROYALTY_TIER_SIZE + // royalty tiers
    1; // bump

  constructor(
    private _data: RegisteredCurrencyAccount,
    private _key: PublicKey,
  ) {}

  get data(): Readonly<RegisteredCurrencyAccount> {
    return this._data;
  }

  get key(): PublicKey {
    return this._key;
  }

  /**
   * Find the `RegisteredCurrency` account address
   * @param program - GalacticMarketplace Program
   * @param currencyMint - the current mint
   * @returns PDA and bump
   */
  static findAddress(
    program: GalacticMarketplaceIDLProgram,
    currencyMint: PublicKey,
  ): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('registered-currency'), currencyMint.toBuffer()],
      program.programId,
    );
  }

  static decodeData(
    account: KeyedAccountInfo,
    program: GalacticMarketplaceIDLProgram,
  ): DecodedAccountData<RegisteredCurrency> {
    return decodeAccount(account, program, RegisteredCurrency);
  }
}
