// @staratlas/galactic-marketplace/tests/galacticMarketplace.test.ts
import {
  AddressLookupTableAccount,
  Connection,
  Finality,
  Keypair,
  LAMPORTS_PER_SOL,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
  TransactionSignature,
} from '@solana/web3.js';
import { findProgramAddressSync } from '@staratlas/anchor/dist/cjs/utils/pubkey';
import {
  AtlasStakingProgram,
  initializeStaking,
  registerStake,
  RegisteredStake,
  StakingVars,
  UserStake,
} from '@staratlas/atlas-staking';
import {
  AsyncSigner,
  InstructionReturn,
  buildDynamicTransactions,
  createAssociatedTokenAccount,
  createMint,
  ixToIxReturn,
  keypairToAsyncSigner,
  mintToTokenAccount,
  readFromRPCNullable,
  readFromRPCOrError,
  sendTransaction,
} from '@staratlas/data-source';
import {
  GalacticMarketplaceProgram,
  MarketVars,
  NOOP_IX_DATA,
  Order,
  cancelOrderWithConnection,
  exchangeOrderWithConnection,
  initializeMarket,
  initializeOrderWithConnection,
  marketVarsDataEquals,
  noop,
  registerCurrency,
} from '../src';
import {
  createInitializeMintInstruction,
  createInitializeTransferHookInstruction,
  createMintToInstruction,
  ExtensionType,
  getAssociatedTokenAddressSync,
  getMintLen,
  TOKEN_2022_PROGRAM_ID,
} from '@solana/spl-token';
import { BN } from 'bn.js';

/**
 * Send many transactions to the chain
 * @param instructions - instruction or instructions
 * @param signer - the transaction signer
 * @param connection - Solana connection object
 * @param commitment - the Solana commitement level
 * @param skipPreflight - whether to skip preflight checks or not
 * @param lookupTables - Optional list of lookup tables to try to build transactions with.
 * @returns array of `TransactionSignature`
 */
export const sendManyToChain = async (
  instructions: InstructionReturn | InstructionReturn[],
  signer: AsyncSigner,
  connection: Connection,
  commitment: Finality = 'confirmed',
  skipPreflight = true,
  lookupTables: AddressLookupTableAccount[] = [],
): Promise<TransactionSignature[]> => {
  const txs = await buildDynamicTransactions(
    instructions,
    signer,
    {
      connection,
    },
    [], // beforeIxs
    [], // afterIxs
    lookupTables,
  );

  if (txs.isErr()) {
    throw txs.error;
  }

  const txSignatures: TransactionSignature[] = [];

  for (const tx of txs.value) {
    const result = await sendTransaction(tx, connection, {
      commitment,
      sendOptions: {
        skipPreflight,
      },
    });

    if (result.value.isErr()) {
      throw result.value.error;
    }

    txSignatures.push(result.value.value);
  }

  return txSignatures;
};

describe('marketplace tests', function () {
  const connection = new Connection('http://localhost:8899', 'confirmed');
  const program = GalacticMarketplaceProgram.buildProgram(
    new PublicKey('MKTdinmdjJV3cnxMMeCGCH6B4o6bvHHdgyv2j3C1Uab'),
    { connection },
  );
  const stakingProgram = AtlasStakingProgram.buildProgram(
    new PublicKey('ASTKsie4j1D6eQHptiGbz6z2jfWAbv5QLETosr6XR4oj'),
    { connection },
  );
  const TransferHookProgramId = new PublicKey(
    'hooKwBRKyzBqxVZFQVpLMKGexhmc6ZNaRAbwWi8uMok',
  );
  const updateAuthorityAccount = keypairToAsyncSigner(
    Keypair.fromSecretKey(
      new Uint8Array([
        170, 92, 226, 151, 121, 147, 2, 238, 53, 50, 158, 46, 255, 232, 52, 223,
        53, 95, 35, 153, 209, 185, 175, 223, 185, 199, 71, 28, 46, 218, 138, 59,
        228, 39, 253, 39, 89, 115, 50, 34, 28, 185, 153, 126, 28, 93, 155, 116,
        204, 216, 78, 18, 88, 98, 149, 138, 47, 140, 35, 87, 87, 145, 79, 186,
      ]),
    ),
  );
  const hunter2 = keypairToAsyncSigner(Keypair.generate());
  const hunter3 = keypairToAsyncSigner(Keypair.generate());

  const initializeExtraAccountMetaList = (mint: PublicKey) => {
    const [extraAccountMetaListPDA, _bump] = PublicKey.findProgramAddressSync(
      [Buffer.from('extra-account-metas'), mint.toBuffer()],
      TransferHookProgramId,
    );
    return new TransactionInstruction({
      programId: TransferHookProgramId,
      keys: [
        {
          pubkey: updateAuthorityAccount.publicKey(),
          isSigner: true,
          isWritable: true,
        }, // Payer
        { pubkey: extraAccountMetaListPDA, isSigner: false, isWritable: true },
        { pubkey: mint, isSigner: false, isWritable: false },
        { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
      ],
      data: Buffer.from([
        92, 197, 174, 197, 41, 124, 19, 3,
      ]) /** hash of 'global:initialize_extra_account_meta_list' */,
    });
  };

  beforeAll(async () => {
    const airdrop = await connection.requestAirdrop(
      updateAuthorityAccount.publicKey(),
      500_000_000_000,
    );
    await connection.confirmTransaction(airdrop);
    const airdrop2 = await connection.requestAirdrop(
      hunter2.publicKey(),
      10_000_000_000,
    );
    await connection.confirmTransaction(airdrop2);
    const airdrop3 = await connection.requestAirdrop(
      hunter3.publicKey(),
      10_000_000_000,
    );
    await connection.confirmTransaction(airdrop3);
  });

  it('should initialize the marketplace', async function () {
    const marketVarsAccount = findProgramAddressSync(
      [Buffer.from('market-vars')],
      program.programId,
    );

    const preRead = await readFromRPCNullable(
      connection,
      program,
      marketVarsAccount[0],
      MarketVars,
    );

    if (preRead === null) {
      await sendManyToChain(
        [initializeMarket(program, updateAuthorityAccount)],
        updateAuthorityAccount,
        connection,
      );

      const marketVars = await readFromRPCOrError(
        connection,
        program,
        marketVarsAccount[0],
        MarketVars,
      );

      expect(
        marketVarsDataEquals(marketVars.data, {
          updateAuthorityMaster: updateAuthorityAccount.publicKey(),
          bump: marketVarsAccount[1],
        }),
      );
    } else {
      console.log('market vars already initialized');
    }
  });

  it('can noop', async function () {
    const noopIx = noop(program);
    // eslint-disable-next-line require-await
    const manualNoop: InstructionReturn = async () => {
      return {
        instruction: new TransactionInstruction({
          keys: [],
          data: NOOP_IX_DATA,
          programId: program.programId,
        }),
        signers: [],
      };
    };

    await sendManyToChain(
      [noopIx, manualNoop],
      updateAuthorityAccount,
      connection,
    );
  });

  it('can place & process an order', async function () {
    const currencyMint = keypairToAsyncSigner(Keypair.generate());
    const tokenAssetMint = keypairToAsyncSigner(Keypair.generate());
    const token2022AssetMint = keypairToAsyncSigner(Keypair.generate());

    // create token2022 mint
    const extensions = [ExtensionType.TransferHook];
    const mintLen = getMintLen(extensions);
    const transferHookProgramId = TransferHookProgramId;
    const mintLamports = await connection.getMinimumBalanceForRentExemption(
      mintLen,
    );
    const token2022MintInstructions = [
      SystemProgram.createAccount({
        fromPubkey: updateAuthorityAccount.publicKey(),
        newAccountPubkey: token2022AssetMint.publicKey(),
        space: mintLen,
        lamports: mintLamports,
        programId: TOKEN_2022_PROGRAM_ID,
      }),
      createInitializeTransferHookInstruction(
        token2022AssetMint.publicKey(),
        updateAuthorityAccount.publicKey(),
        transferHookProgramId,
        TOKEN_2022_PROGRAM_ID,
      ),
      createInitializeMintInstruction(
        token2022AssetMint.publicKey(),
        0,
        updateAuthorityAccount.publicKey(),
        null,
        TOKEN_2022_PROGRAM_ID,
      ),
    ];

    const royaltyRate = 500;
    const baseAmount = 100;
    const price = new BN(1.5 * LAMPORTS_PER_SOL);
    const originationQty = 2;
    const registeredStake = RegisteredStake.findAddress(
      stakingProgram,
      updateAuthorityAccount.publicKey(),
      currencyMint.publicKey(),
      currencyMint.publicKey(),
    )[0];

    // setup
    const saVault = getAssociatedTokenAddressSync(
      currencyMint.publicKey(),
      updateAuthorityAccount.publicKey(),
      true,
    );

    const stakingVars = await readFromRPCNullable(
      connection,
      stakingProgram,
      StakingVars.findAddress(stakingProgram)[0],
      StakingVars,
    );

    await sendManyToChain(
      [
        createMint(currencyMint, 9, updateAuthorityAccount.publicKey(), null),
        createAssociatedTokenAccount(
          currencyMint.publicKey(),
          updateAuthorityAccount.publicKey(),
        ).instructions,
        registerCurrency(
          program,
          updateAuthorityAccount,
          currencyMint.publicKey(),
          royaltyRate,
          saVault,
        ),
        ...(stakingVars == null
          ? [initializeStaking(stakingProgram, updateAuthorityAccount)]
          : []),
        registerStake(
          stakingProgram,
          updateAuthorityAccount,
          currencyMint.publicKey(),
          currencyMint.publicKey(),
          1,
          3600,
        ),
        createMint(tokenAssetMint, 0, updateAuthorityAccount.publicKey(), null),
        ...token2022MintInstructions.map((it, idx) => {
          if (idx === 0) {
            return ixToIxReturn(it, [token2022AssetMint]);
          }
          return ixToIxReturn(it);
        }),
        ixToIxReturn(
          initializeExtraAccountMetaList(token2022AssetMint.publicKey()),
          [updateAuthorityAccount],
        ),
        createAssociatedTokenAccount(
          currencyMint.publicKey(),
          hunter2.publicKey(),
        ).instructions,
        mintToTokenAccount(
          updateAuthorityAccount,
          currencyMint.publicKey(),
          getAssociatedTokenAddressSync(
            currencyMint.publicKey(),
            hunter2.publicKey(),
            true,
          ),
          baseAmount * LAMPORTS_PER_SOL,
        ),
        createAssociatedTokenAccount(
          tokenAssetMint.publicKey(),
          hunter2.publicKey(),
        ).instructions,
        mintToTokenAccount(
          updateAuthorityAccount,
          tokenAssetMint.publicKey(),
          getAssociatedTokenAddressSync(
            tokenAssetMint.publicKey(),
            hunter2.publicKey(),
            true,
          ),
          baseAmount,
        ),
        createAssociatedTokenAccount(
          token2022AssetMint.publicKey(),
          hunter2.publicKey(),
          true,
          TOKEN_2022_PROGRAM_ID,
        ).instructions,
        ixToIxReturn(
          createMintToInstruction(
            token2022AssetMint.publicKey(),
            getAssociatedTokenAddressSync(
              token2022AssetMint.publicKey(),
              hunter2.publicKey(),
              true,
              TOKEN_2022_PROGRAM_ID,
            ),
            updateAuthorityAccount.publicKey(),
            baseAmount,
            [],
            TOKEN_2022_PROGRAM_ID,
          ),
          [updateAuthorityAccount],
        ),
        createAssociatedTokenAccount(
          currencyMint.publicKey(),
          hunter3.publicKey(),
        ).instructions,
        mintToTokenAccount(
          updateAuthorityAccount,
          currencyMint.publicKey(),
          getAssociatedTokenAddressSync(
            currencyMint.publicKey(),
            hunter3.publicKey(),
            true,
          ),
          baseAmount * LAMPORTS_PER_SOL,
        ),
        createAssociatedTokenAccount(
          tokenAssetMint.publicKey(),
          hunter3.publicKey(),
        ).instructions,
        mintToTokenAccount(
          updateAuthorityAccount,
          tokenAssetMint.publicKey(),
          getAssociatedTokenAddressSync(
            tokenAssetMint.publicKey(),
            hunter3.publicKey(),
            true,
          ),
          baseAmount,
        ),
        createAssociatedTokenAccount(
          token2022AssetMint.publicKey(),
          hunter3.publicKey(),
          true,
          TOKEN_2022_PROGRAM_ID,
        ).instructions,
        ixToIxReturn(
          createMintToInstruction(
            token2022AssetMint.publicKey(),
            getAssociatedTokenAddressSync(
              token2022AssetMint.publicKey(),
              hunter3.publicKey(),
              true,
              TOKEN_2022_PROGRAM_ID,
            ),
            updateAuthorityAccount.publicKey(),
            baseAmount,
            [],
            TOKEN_2022_PROGRAM_ID,
          ),
          [updateAuthorityAccount],
        ),
      ],
      updateAuthorityAccount,
      connection,
    );

    // create orders
    const order1 = await initializeOrderWithConnection(
      connection,
      'buy',
      hunter2,
      price,
      originationQty,
      tokenAssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
    );
    const order2 = await initializeOrderWithConnection(
      connection,
      'sell',
      hunter2,
      price,
      originationQty,
      tokenAssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
    );
    await sendManyToChain(
      [...order1.instructions, ...order2.instructions],
      updateAuthorityAccount,
      connection,
    );

    const order1Account = await readFromRPCOrError(
      connection,
      program,
      order1.orderAccount,
      Order,
    );
    expect(
      order1Account.data.orderInitializerPubkey.equals(hunter2.publicKey()),
    ).toBe(true);
    expect(
      order1Account.data.currencyMint.equals(currencyMint.publicKey()),
    ).toBe(true);
    expect(
      order1Account.data.assetMint.equals(tokenAssetMint.publicKey()),
    ).toBe(true);
    expect(
      order1Account.data.initializerCurrencyTokenAccount.equals(
        getAssociatedTokenAddressSync(
          currencyMint.publicKey(),
          hunter2.publicKey(),
          true,
        ),
      ),
    ).toBe(true);
    expect(
      order1Account.data.initializerAssetTokenAccount.equals(
        getAssociatedTokenAddressSync(
          tokenAssetMint.publicKey(),
          hunter2.publicKey(),
          true,
        ),
      ),
    ).toBe(true);
    expect(order1Account.data.orderOriginationQty.toNumber()).toBe(
      originationQty,
    );
    expect(order1Account.data.price.toNumber()).toBe(price.toNumber());

    // do the exchanges
    const order2Account = await readFromRPCOrError(
      connection,
      program,
      order2.orderAccount,
      Order,
    );
    const exchange1 = await exchangeOrderWithConnection(
      connection,
      order1Account,
      hunter3,
      getAssociatedTokenAddressSync(
        tokenAssetMint.publicKey(),
        hunter3.publicKey(),
        true,
      ),
      originationQty,
      price,
      saVault,
      stakingProgram.programId,
      registeredStake,
      UserStake.findAddress(
        stakingProgram,
        hunter3.publicKey(),
        registeredStake,
      )[0] /** seller */,
      program,
      [],
    );
    const exchange2 = await exchangeOrderWithConnection(
      connection,
      order2Account,
      hunter3,
      getAssociatedTokenAddressSync(
        currencyMint.publicKey(),
        hunter3.publicKey(),
        true,
      ),
      originationQty,
      price,
      saVault,
      stakingProgram.programId,
      registeredStake,
      UserStake.findAddress(
        stakingProgram,
        hunter2.publicKey(),
        registeredStake,
      )[0] /** seller */,
      program,
      [],
    );
    await sendManyToChain(
      [...exchange1, ...exchange2],
      updateAuthorityAccount,
      connection,
    );

    // token 2022
    const order3 = await initializeOrderWithConnection(
      connection,
      'buy',
      hunter2,
      price,
      originationQty,
      token2022AssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
      TransferHookProgramId,
    );
    const order4 = await initializeOrderWithConnection(
      connection,
      'sell',
      hunter2,
      price,
      originationQty,
      token2022AssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
      TransferHookProgramId,
    );
    await sendManyToChain(
      [...order3.instructions, ...order4.instructions],
      updateAuthorityAccount,
      connection,
    );
    const order3Account = await readFromRPCOrError(
      connection,
      program,
      order3.orderAccount,
      Order,
    );
    const order4Account = await readFromRPCOrError(
      connection,
      program,
      order4.orderAccount,
      Order,
    );
    const exchange3 = await exchangeOrderWithConnection(
      connection,
      order3Account,
      hunter3,
      getAssociatedTokenAddressSync(
        token2022AssetMint.publicKey(),
        hunter3.publicKey(),
        true,
        TOKEN_2022_PROGRAM_ID,
      ),
      originationQty,
      price,
      saVault,
      stakingProgram.programId,
      registeredStake,
      UserStake.findAddress(
        stakingProgram,
        hunter3.publicKey(),
        registeredStake,
      )[0] /** seller */,
      program,
      [],
      TransferHookProgramId,
    );
    const exchange4 = await exchangeOrderWithConnection(
      connection,
      order4Account,
      hunter3,
      getAssociatedTokenAddressSync(
        currencyMint.publicKey(),
        hunter3.publicKey(),
        true,
      ),
      originationQty,
      price,
      saVault,
      stakingProgram.programId,
      registeredStake,
      UserStake.findAddress(
        stakingProgram,
        hunter2.publicKey(),
        registeredStake,
      )[0] /** seller */,
      program,
      [],
      TransferHookProgramId,
    );
    await sendManyToChain(
      [...exchange3, ...exchange4],
      updateAuthorityAccount,
      connection,
    );

    // cancel orders
    const order5 = await initializeOrderWithConnection(
      connection,
      'buy',
      hunter3,
      price,
      originationQty,
      tokenAssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
    );
    const order6 = await initializeOrderWithConnection(
      connection,
      'sell',
      hunter3,
      price,
      originationQty,
      tokenAssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
    );
    const order7 = await initializeOrderWithConnection(
      connection,
      'buy',
      hunter2,
      price,
      originationQty,
      token2022AssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
      TransferHookProgramId,
    );
    const order8 = await initializeOrderWithConnection(
      connection,
      'sell',
      hunter2,
      price,
      originationQty,
      token2022AssetMint.publicKey(),
      currencyMint.publicKey(),
      program,
      TransferHookProgramId,
    );
    await sendManyToChain(
      [
        ...order5.instructions,
        ...order6.instructions,
        ...order7.instructions,
        ...order8.instructions,
      ],
      updateAuthorityAccount,
      connection,
    );
    const order5Account = await readFromRPCOrError(
      connection,
      program,
      order5.orderAccount,
      Order,
    );
    const order6Account = await readFromRPCOrError(
      connection,
      program,
      order6.orderAccount,
      Order,
    );
    const order7Account = await readFromRPCOrError(
      connection,
      program,
      order7.orderAccount,
      Order,
    );
    const order8Account = await readFromRPCOrError(
      connection,
      program,
      order8.orderAccount,
      Order,
    );
    const cancel5 = await cancelOrderWithConnection(
      connection,
      hunter3.publicKey(),
      order5Account,
      program,
    );
    const cancel6 = await cancelOrderWithConnection(
      connection,
      hunter3.publicKey(),
      order6Account,
      program,
    );
    const cancel7 = await cancelOrderWithConnection(
      connection,
      hunter2.publicKey(),
      order7Account,
      program,
      TransferHookProgramId,
    );
    const cancel8 = await cancelOrderWithConnection(
      connection,
      hunter2.publicKey(),
      order8Account,
      program,
      TransferHookProgramId,
    );
    await sendManyToChain(
      [...cancel5, ...cancel6, ...cancel7, ...cancel8],
      updateAuthorityAccount,
      connection,
    );
  });
});
