// @staratlas/galactic-marketplace/dist/src/instructions/addFeeExemption.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addFeeExpemption = void 0;
const web3_js_1 = require("@solana/web3.js");
const anchor_1 = require("@staratlas/anchor");
const __1 = require("..");
const feeReduction_1 = require("../state/feeReduction");
/**
 * Returns an instruction which flags an account as fee exempt. Update authority must match marketVarsAccount.
 * @param program - Marketplace program
 * @param updateAuthorityAccount - Public key of registered update authority
 * @param feeExemptTarget - Target account to exempt from fees
 * @returns InstructionReturn
 */
function addFeeExpemption(program, updateAuthorityAccount, feeExemptTarget) {
    const discount = new anchor_1.BN(1000000);
    const feeExemptAccount = feeReduction_1.FeeReduction.findAddress(feeExemptTarget, program)[0];
    const marketVarsAccount = __1.MarketVars.findAddress(program)[0];
    return (funder) => __awaiter(this, void 0, void 0, function* () {
        return [
            {
                instruction: yield program.methods
                    .addFeeExemption(discount)
                    .accountsStrict({
                    updateAuthorityMaster: updateAuthorityAccount.publicKey(),
                    funder: funder.publicKey(),
                    feeExemptTarget,
                    feeExemptAccount,
                    marketVarsAccount,
                    systemProgram: web3_js_1.SystemProgram.programId,
                })
                    .instruction(),
                signers: [updateAuthorityAccount, funder],
            },
        ];
    });
}
exports.addFeeExpemption = addFeeExpemption;
