// @staratlas/galactic-marketplace/dist/src/instructions/createOpenOrdersCounter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOpenOrdersCounter = void 0;
const web3_js_1 = require("@solana/web3.js");
const state_1 = require("../state");
/**
 * Creates OpenOrdersCounter Account
 * @param program - Galactic Marketplace Program
 * @param user - User Key
 * @param depositMint - Deposite Mint key
 * @returns InstructionReturn
 */
function createOpenOrdersCounter(program, user, depositMint) {
    return (funder) => __awaiter(this, void 0, void 0, function* () {
        return ({
            instruction: yield program.methods
                .initializeOpenOrdersCounter()
                .accountsStrict({
                payer: funder.publicKey(),
                user,
                openOrdersCounter: state_1.OpenOrdersCounter.findAddress(program, user, depositMint)[0],
                depositMint,
                systemProgram: web3_js_1.SystemProgram.programId,
            })
                .instruction(),
            signers: [funder],
        });
    });
}
exports.createOpenOrdersCounter = createOpenOrdersCounter;
