// @staratlas/galactic-marketplace/dist/src/instructions/updateAtlasRate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAtlasRate = void 0;
const web3_js_1 = require("@solana/web3.js");
const bn_js_1 = __importDefault(require("bn.js"));
const state_1 = require("../state");
const atlasRateAccount_1 = require("../state/atlasRateAccount");
/**
 * `atlas_rate` * Atlas sub-tokens = USDC sub-tokens
 * `atlas_rate` = `1`/(`100` * Atlas price in USD)
 * @param program - The GalacticMarketplace program
 * @param updateAuthorityAccount - The authority of the market
 * @param rate - The rate of Atlas to USDC
 * @returns InstructionReturn
 */
function updateAtlasRate(program, updateAuthorityAccount, rate) {
    return (funder) => __awaiter(this, void 0, void 0, function* () {
        return ({
            instruction: yield program.methods
                .updateAtlasRate(new bn_js_1.default(Math.floor(rate * 1000000000)))
                .accountsStrict({
                funder: funder.publicKey(),
                updateAuthorityAccount: updateAuthorityAccount.publicKey(),
                atlasRate: atlasRateAccount_1.AtlasRate.findAddress(program)[0],
                marketVarsAccount: state_1.MarketVars.findAddress(program)[0],
                systemProgram: web3_js_1.SystemProgram.programId,
            })
                .instruction(),
            signers: [updateAuthorityAccount, funder],
        });
    });
}
exports.updateAtlasRate = updateAtlasRate;
